VM: Install python packages fails on uninstall pexpect

Trying to create a new VM.

For whatever reason
https://github.com/SyneRBI/SIRF-SuperBuild/blob/ec4b9c50bb650101e232590405c06eaf417256a6/VirtualBox/scripts/INSTALL_python_packages.sh#L70
triggers an uninstall of `pexpect` with `pip`, but this fails as there's one installed by `apt` I guess.

screenshot (sorry, no copy-paste capability on this VM yet)
 ![image](https://user-images.githubusercontent.com/6362141/178552937-623f0742-a178-4a10-ab99-e14aaed95cae.png)

The installed package is at `/usr/lib/python3/dist-packages/pexpect`. Maybe it got there via https://github.com/SyneRBI/SIRF-SuperBuild/blob/d9de45ba5163eb175d6bb8cd37a2031755fd4d26/VirtualBox/scripts/INSTALL_prerequisites_with_apt-get.sh#L39-L40

OK, I have now passed the install prerequisites step! 
Next problem in line:

```
CMake Error at SuperBuild/External_astra-toolbox.cmake:32 (message):
    default:   astra-toolkit binding to Python depends on Cython
```

And as [mentioned ](https://github.com/SyneRBI/SIRF-SuperBuild/issues/742#issuecomment-1205226105) I need to make the find Cython accept hints.

After installation of all dependencies the build gets to find Cython but it is not found.

Notice that I try on the default location `/usr/local/bin/cython` and on what I think it should be for a pip install for the user, `${HOME}/.local/bin`. I am not sure this is right, however, since the install is done by vagrant with the user vagrant (I think?)

```
--
 Looking for Cython among /usr/bin/cython3;/home/sirfuser/.local/bin/cython
    default: statusTrying with /usr/bin/cython3
    default: statusTrying with /home/sirfuser/.local/bin/cython
    default: CMake Warning at CMake/FindCython.cmake:54 (message):
    default:   Cython not found!
    default: Call Stack (most recent call first):
    default:   SuperBuild/External_astra-toolbox.cmake:30 (find_package)
    default:   CMake/ExternalProjectDependency.cmake:593 (include)
    default:   SuperBuild/External_astra-python-wrapper.cmake:29 (ExternalProject_Include_Dependencies)
    default:   CMake/ExternalProjectDependency.cmake:593 (include)
    default:   SuperBuild/External_CIL-ASTRA.cmake:28 (ExternalProject_Include_Dependencies)
    default:   CMake/ExternalProjectDependency.cmake:593 (include)
    default:   CMake/ExternalProjectDependency.cmake:639 (ExternalProject_Include_Dependencies)
    default:   SuperBuild.cmake:329 (ExternalProject_Include_Dependencies)
    default:   CMakeLists.txt:42 (include)
    default:
    default:
    default: CMake Error at SuperBuild/External_astra-toolbox.cmake:32 (message):
    default:   astra-toolkit binding to Python depends on Cython
```

I tried to use `sudo -u sirfuser python -m pip install -U -r requirements.txt` but it still failed (I'll try again)

```
default: python3: can't open file 'sirfuser': [Errno 2] No such file or directory
    default: An error occurred in /home/sirfuser/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.sh at line 76. Current working-dir: /home/sirfuser/devel/SIRF-SuperBuild
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,15 @@
 # ChangeLog
-# v3.3.0
+## v3.3.1
+- VM: 
+   - Updates to run using docker scripts
+   - installs custom pip and all python prerequisites with pip
+   - Bugfix in finding cython and python in UPDATE.sh
+- CMake:
+   - FindCython allows hints
+- updated versions:
+   - numba < 0.51, due to incompatibilities with python 3.6 (which is installed in the GHA and VM)
+   
+## v3.3.0
 - known problems:
    - VM and jupyterhub scripts need merging various fixes
 - gemeric *.cmake fixes:

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -22,8 +22,9 @@ message(STATUS "CYTHON_EXECUTABLES ${CYTHON_EXECUTABLES}")
 
 set(Cython_FOUND FALSE)
 foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
+    message(STATUS "Trying with ${cy_exe}!")
     IF (cy_exe)
-        message(STATUS "Trying with ${cy_exe}")
+        
         # Try to run Cython, to make sure it works:
         execute_process(
             COMMAND ${cy_exe} "--version"

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -12,8 +12,8 @@ SET(CYTHON_EXECUTABLE cython CACHE STRING "Cython executable name")
 SET(CYTHON_FLAGS --cplus --fast-fail)
 
 set(SEARCH_CYTHON_PATH "")
-if (CYTHON_PATH)
-  set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE};${CYTHON_PATH}/cython")
+if (CYTHON_ROOT)
+  set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE};${CYTHON_ROOT}/cython")
 else()
 set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE}")
 endif()
@@ -21,6 +21,7 @@ endif()
 set(Cython_FOUND FALSE)
 foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
     IF (cy_exe)
+        message(STATUS "Trying with ${cy_exe}")
         # Try to run Cython, to make sure it works:
         execute_process(
             COMMAND ${cy_exe} "--version"
@@ -32,9 +33,11 @@ foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
             # Only if cython exits with the return code 0, we know that all is ok:
             SET(Cython_FOUND TRUE)
             SET(Cython_Compilation_Failed FALSE)
+            message(STATUS "${cy_exe} works"
             break()
         else (CYTHON_RESULT EQUAL 0)
             SET(Cython_Compilation_Failed TRUE)
+            message(STATUS "${cy_exe} does not work")
         endif (CYTHON_RESULT EQUAL 0)
     endif (cy_exe)
 endforeach()

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -11,33 +11,46 @@
 SET(CYTHON_EXECUTABLE cython CACHE STRING "Cython executable name")
 SET(CYTHON_FLAGS --cplus --fast-fail)
 
-SET(Cython_FOUND FALSE)
-IF (CYTHON_EXECUTABLE)
-    # Try to run Cython, to make sure it works:
-    execute_process(
-        COMMAND ${CYTHON_EXECUTABLE} "--version"
-        RESULT_VARIABLE CYTHON_RESULT
-        OUTPUT_QUIET
-        ERROR_QUIET
-        )
-    if (CYTHON_RESULT EQUAL 0)
-        # Only if cython exits with the return code 0, we know that all is ok:
-        SET(Cython_FOUND TRUE)
-        SET(Cython_Compilation_Failed FALSE)
-    else (CYTHON_RESULT EQUAL 0)
-        SET(Cython_Compilation_Failed TRUE)
-    endif (CYTHON_RESULT EQUAL 0)
-ENDIF (CYTHON_EXECUTABLE)
+set(SEARCH_CYTHON_PATH "")
+if (CYTHON_PATH)
+  set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE};${CYTHON_PATH}/cython")
+else()
+set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE}")
+endif()
+
+set(Cython_FOUND FALSE)
+foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
+    IF (cy_exe)
+        # Try to run Cython, to make sure it works:
+        execute_process(
+            COMMAND ${cy_exe} "--version"
+            RESULT_VARIABLE CYTHON_RESULT
+            OUTPUT_QUIET
+            ERROR_QUIET
+            )
+        if (CYTHON_RESULT EQUAL 0)
+            # Only if cython exits with the return code 0, we know that all is ok:
+            SET(Cython_FOUND TRUE)
+            SET(Cython_Compilation_Failed FALSE)
+            break()
+        else (CYTHON_RESULT EQUAL 0)
+            SET(Cython_Compilation_Failed TRUE)
+        endif (CYTHON_RESULT EQUAL 0)
+    endif (cy_exe)
+endforeach()
+
+
 
 
 IF (Cython_FOUND)
 	IF (NOT Cython_FIND_QUIETLY)
-		MESSAGE(STATUS "Found CYTHON: ${CYTHON_EXECUTABLE}")
+		MESSAGE(STATUS "Found CYTHON version ${OUTPUT_QUIET}: ${CYTHON_EXECUTABLE}")
 	ENDIF (NOT Cython_FIND_QUIETLY)
 ELSE (Cython_FOUND)
-	IF (Cython_FIND_REQUIRED)
+    message(WARNING "Cython not found!")
+	if (Cython_FIND_REQUIRED)
         if(Cython_Compilation_Failed)
-            MESSAGE(STATUS "Found CYTHON: ${CYTHON_EXECUTABLE}")
+            MESSAGE(STATUS "Found CYTHON: ${CYTHON_EXECUTABLE} But compilation failed")
 			# On Win the testing of Cython does not return any accessible value, so the test is not carried out. Fresh Cython install was tested and works.
 			IF(NOT MSVC)
 				MESSAGE(FATAL_ERROR "Your Cython version is too old. Please upgrade Cython.")
@@ -45,6 +58,6 @@ ELSE (Cython_FOUND)
         else(Cython_Compilation_Failed)
             MESSAGE(FATAL_ERROR "Could not find Cython. Please install Cython.")
         endif(Cython_Compilation_Failed)
-	ENDIF (Cython_FIND_REQUIRED)
-ENDIF (Cython_FOUND)
+	endif (Cython_FIND_REQUIRED)
+endif (Cython_FOUND)
 

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -22,23 +22,26 @@ message(STATUS "Looking for Cython among ${CYTHON_EXECUTABLES}")
 
 set(Cython_FOUND FALSE)
 foreach(CYTHON_EXECUTABLE IN LISTS CYTHON_EXECUTABLES)
-    IF (CYTHON_EXECUTABLE)
+    message(status "Trying with ${cy_exe}")
+    IF (cy_exe)
         # Try to run Cython, to make sure it works:
         execute_process(
-            COMMAND ${CYTHON_EXECUTABLE} "--version"
+            COMMAND ${cy_exe} "--version"
             RESULT_VARIABLE CYTHON_RESULT
-            OUTPUT_QUIET
+            OUTPUT_VARIABLE cython_version
             ERROR_QUIET
             )
         if (CYTHON_RESULT EQUAL 0)
             # Only if cython exits with the return code 0, we know that all is ok:
             SET(Cython_FOUND TRUE)
             SET(Cython_Compilation_Failed FALSE)
+            set(CYTHON_EXECUTABLE ${cy_exe})
+            message(status "YAS! ${cy_exe} ${CYTHON_EXECUTABLE} ${cython_version}")
             break()
         else (CYTHON_RESULT EQUAL 0)
             SET(Cython_Compilation_Failed TRUE)
         endif (CYTHON_RESULT EQUAL 0)
-    endif (CYTHON_EXECUTABLE)
+    endif (cy_exe)
 endforeach()
 
 
@@ -46,7 +49,7 @@ endforeach()
 
 IF (Cython_FOUND)
 	IF (NOT Cython_FIND_QUIETLY)
-		MESSAGE(STATUS "Found CYTHON version ${OUTPUT_QUIET}: ${CYTHON_EXECUTABLE}")
+		MESSAGE(STATUS "Found CYTHON version ${cython_version}: ${CYTHON_EXECUTABLE}")
 	ENDIF (NOT Cython_FIND_QUIETLY)
 ELSE (Cython_FOUND)
     message(WARNING "Cython not found!")

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -18,16 +18,14 @@ else()
   set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE}")
 endif()
 
-message(STATUS "CYTHON_EXECUTABLES ${CYTHON_EXECUTABLES}")
+message(STATUS "Looking for Cython among ${CYTHON_EXECUTABLES}")
 
 set(Cython_FOUND FALSE)
-foreach(cy_exe IN LISTS CYTHON_EXECUTABLES)
-    message(STATUS "Trying with ${cy_exe}!")
-    IF (cy_exe)
-        
+foreach(CYTHON_EXECUTABLE IN LISTS CYTHON_EXECUTABLES)
+    IF (CYTHON_EXECUTABLE)
         # Try to run Cython, to make sure it works:
         execute_process(
-            COMMAND ${cy_exe} "--version"
+            COMMAND ${CYTHON_EXECUTABLE} "--version"
             RESULT_VARIABLE CYTHON_RESULT
             OUTPUT_QUIET
             ERROR_QUIET
@@ -36,13 +34,11 @@ foreach(cy_exe IN LISTS CYTHON_EXECUTABLES)
             # Only if cython exits with the return code 0, we know that all is ok:
             SET(Cython_FOUND TRUE)
             SET(Cython_Compilation_Failed FALSE)
-            message(STATUS "${cy_exe} works")
             break()
         else (CYTHON_RESULT EQUAL 0)
             SET(Cython_Compilation_Failed TRUE)
-            message(STATUS "${cy_exe} does not work")
         endif (CYTHON_RESULT EQUAL 0)
-    endif (cy_exe)
+    endif (CYTHON_EXECUTABLE)
 endforeach()
 
 

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -18,6 +18,8 @@ else()
 set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE}")
 endif()
 
+message(STATUS "CYTHON_EXECUTABLES ${CYTHON_EXECUTABLES}")
+
 set(Cython_FOUND FALSE)
 foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
     IF (cy_exe)

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -21,7 +21,7 @@ endif()
 message(STATUS "Looking for Cython among ${CYTHON_EXECUTABLES}")
 
 set(Cython_FOUND FALSE)
-foreach(CYTHON_EXECUTABLE IN LISTS CYTHON_EXECUTABLES)
+foreach(cy_exe IN LISTS CYTHON_EXECUTABLES)
     message(status "Trying with ${cy_exe}")
     IF (cy_exe)
         # Try to run Cython, to make sure it works:

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -29,7 +29,7 @@ foreach(cy_exe IN LISTS CYTHON_EXECUTABLES)
             COMMAND ${cy_exe} "--version"
             RESULT_VARIABLE CYTHON_RESULT
             OUTPUT_QUIET 
-            ERROR_QUIET cython_version
+            ERROR_VARIABLE cython_version
             )
         if (CYTHON_RESULT EQUAL 0)
             # Only if cython exits with the return code 0, we know that all is ok:

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -28,8 +28,8 @@ foreach(cy_exe IN LISTS CYTHON_EXECUTABLES)
         execute_process(
             COMMAND ${cy_exe} "--version"
             RESULT_VARIABLE CYTHON_RESULT
-            OUTPUT_VARIABLE cython_version
-            ERROR_QUIET
+            OUTPUT_QUIET 
+            ERROR_QUIET cython_version
             )
         if (CYTHON_RESULT EQUAL 0)
             # Only if cython exits with the return code 0, we know that all is ok:

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -11,7 +11,6 @@
 SET(CYTHON_EXECUTABLE cython CACHE STRING "Cython executable name")
 SET(CYTHON_FLAGS --cplus --fast-fail)
 
-set(SEARCH_CYTHON_PATH "")
 if (CYTHON_ROOT)
   set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE};${CYTHON_ROOT}/cython")
 else()

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -33,7 +33,7 @@ foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
             # Only if cython exits with the return code 0, we know that all is ok:
             SET(Cython_FOUND TRUE)
             SET(Cython_Compilation_Failed FALSE)
-            message(STATUS "${cy_exe} works"
+            message(STATUS "${cy_exe} works")
             break()
         else (CYTHON_RESULT EQUAL 0)
             SET(Cython_Compilation_Failed TRUE)

--- a/CMake/FindCython.cmake
+++ b/CMake/FindCython.cmake
@@ -15,13 +15,13 @@ set(SEARCH_CYTHON_PATH "")
 if (CYTHON_ROOT)
   set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE};${CYTHON_ROOT}/cython")
 else()
-set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE}")
+  set(CYTHON_EXECUTABLES "${CYTHON_EXECUTABLE}")
 endif()
 
 message(STATUS "CYTHON_EXECUTABLES ${CYTHON_EXECUTABLES}")
 
 set(Cython_FOUND FALSE)
-foreach(cy_exe IN LISTS ${CYTHON_EXECUTABLES})
+foreach(cy_exe IN LISTS CYTHON_EXECUTABLES)
     message(STATUS "Trying with ${cy_exe}!")
     IF (cy_exe)
         

--- a/VirtualBox/scripts/INSTALL_prerequisites_with_apt-get.sh
+++ b/VirtualBox/scripts/INSTALL_prerequisites_with_apt-get.sh
@@ -36,11 +36,11 @@ ${APT_GET_INSTALL} expect
 echo "Installing python APT packages"
 # we will use pip for most
 # some extra package needed for jupyter
-qt=pyqt5
-${APT_GET_INSTALL} python3-dev python3-pip python3-tk python3-${qt} python3-${qt}.qtsvg python3-${qt}.qtwebkit
+# qt=pyqt5
+# ${APT_GET_INSTALL} python3-dev python3-pip python3-tk python3-${qt} python3-${qt}.qtsvg python3-${qt}.qtwebkit
 
-echo "Installing CIL pre-requisites..."
-${APT_GET_INSTALL} cython3 python3-h5py python3-wget
+# echo "Installing CIL pre-requisites..."
+# ${APT_GET_INSTALL} cython3 python3-h5py python3-wget
 # dependency for ASTRA-toolbox autotools-dev automake autogen autoconf libtool
 
 echo "Run INSTALL_python_packages.sh after this."

--- a/VirtualBox/scripts/INSTALL_prerequisites_with_apt-get.sh
+++ b/VirtualBox/scripts/INSTALL_prerequisites_with_apt-get.sh
@@ -14,6 +14,9 @@ SUDO=sudo
 
 # TODO would be better to guarantee absolute path for SCRIPTS
 SCRIPTS="$(dirname $0)/../../docker"
+# uninstall pexpect which is preinstalled in the VM
+# https://github.com/SyneRBI/SIRF-SuperBuild/issues/742#issuecomment-1205090681
+$SUDO apt remove -y python3-pexpect
 
 $SUDO "$SCRIPTS/build_essential-ubuntu.sh"
 

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -72,9 +72,9 @@ ${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev
 # TODO would be better to guarantee absolute path for SCRIPTS
 SCRIPTS="$(dirname $0)/../../docker"
 
-${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
-${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
-${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements-service.txt
+${SUDO} -u sirfuser ${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
+${PYTHON} -u sirfuser -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
+${PYTHON} -u sirfuser -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements-service.txt
 # $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 # $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
 # $PYTHON -m pip install $PIPOPTIONS jupyter spyder

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -68,9 +68,13 @@ done
 export DEBIAN_FRONTEND=noninteractive
 SUDO=sudo
 APT_GET_INSTALL="$SUDO apt-get install -y --no-install-recommends"
-${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev  
+${APT_GET_INSTALL} ${PYTHON}-dev  
 # TODO would be better to guarantee absolute path for SCRIPTS
 SCRIPTS="$(dirname $0)/../../docker"
+# installs pip
+curl $($PYTHON ${SCRIPTS}/get_pip_download_link.py) > get-pip.py
+${PYTHON} get-pip.py
+rm get-pip.py
 
 ${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
 ${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -73,8 +73,8 @@ ${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev
 SCRIPTS="$(dirname $0)/../../docker"
 
 ${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
-${PYTHON} -u sirfuser -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
-${PYTHON} -u sirfuser -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements-service.txt
+${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
+${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements-service.txt
 # $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 # $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
 # $PYTHON -m pip install $PIPOPTIONS jupyter spyder

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -72,7 +72,8 @@ ${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev
 # TODO would be better to guarantee absolute path for SCRIPTS
 SCRIPTS="$(dirname $0)/../../docker"
 
-${PYTHON} -m pip install $PIPOPTIONS -y -U -r ${SCRIPTS}/requirements.txt
+${PYTHON} -m pip install $PIPOPTIONS -U setuptools
+${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
 # $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 # $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
 # $PYTHON -m pip install $PIPOPTIONS jupyter spyder

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -74,6 +74,7 @@ SCRIPTS="$(dirname $0)/../../docker"
 
 ${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
 ${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
+${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements-service.txt
 # $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 # $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
 # $PYTHON -m pip install $PIPOPTIONS jupyter spyder

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -72,7 +72,7 @@ ${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev
 # TODO would be better to guarantee absolute path for SCRIPTS
 SCRIPTS="$(dirname $0)/../../docker"
 
-${SUDO} -u sirfuser ${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
+${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
 ${PYTHON} -u sirfuser -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
 ${PYTHON} -u sirfuser -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements-service.txt
 # $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -65,6 +65,11 @@ while (( "$#" )); do
   esac
 done
 
+export DEBIAN_FRONTEND=noninteractive
+SUDO=sudo
+APT_GET_INSTALL="$SUDO apt-get install -y --no-install-recommends"
+${APT_GET_INSTALL} ${PYTHON}-pip
+
 $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
 $PYTHON -m pip install $PIPOPTIONS jupyter spyder

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -69,13 +69,15 @@ export DEBIAN_FRONTEND=noninteractive
 SUDO=sudo
 APT_GET_INSTALL="$SUDO apt-get install -y --no-install-recommends"
 ${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev  
+# TODO would be better to guarantee absolute path for SCRIPTS
+SCRIPTS="$(dirname $0)/../../docker"
 
-$PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
-$PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
-$PYTHON -m pip install $PIPOPTIONS jupyter spyder
-$PYTHON -m pip uninstall $PIPOPTIONS -y spyder-kernels
+${PYTHON} -m pip install $PIPOPTIONS -y -U -r ${SCRIPTS}/requirements.txt
+# $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
+# $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm
+# $PYTHON -m pip install $PIPOPTIONS jupyter spyder
+# $PYTHON -m pip uninstall $PIPOPTIONS -y spyder-kernels
 
 # otherwise Jupyter uses py 2 even when you choose py3: https://github.com/jupyter/jupyter/issues/270
 $PYTHON -m ipykernel install --user  
-# CIL
-$PYTHON -m pip install $PIPOPTIONS pillow olefile
+

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -72,7 +72,7 @@ ${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev
 # TODO would be better to guarantee absolute path for SCRIPTS
 SCRIPTS="$(dirname $0)/../../docker"
 
-${PYTHON} -m pip install $PIPOPTIONS -U setuptools
+${PYTHON} -m pip install $PIPOPTIONS -U setuptools wheel
 ${PYTHON} -m pip install $PIPOPTIONS -U -r ${SCRIPTS}/requirements.txt
 # $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 # $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm

--- a/VirtualBox/scripts/INSTALL_python_packages.sh
+++ b/VirtualBox/scripts/INSTALL_python_packages.sh
@@ -68,7 +68,7 @@ done
 export DEBIAN_FRONTEND=noninteractive
 SUDO=sudo
 APT_GET_INSTALL="$SUDO apt-get install -y --no-install-recommends"
-${APT_GET_INSTALL} ${PYTHON}-pip
+${APT_GET_INSTALL} ${PYTHON}-pip ${PYTHON}-dev  
 
 $PYTHON -m pip install $PIPOPTIONS --upgrade pip wheel setuptools
 $PYTHON -m pip install $PIPOPTIONS --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib nose coverage docopt deprecation nibabel pytest tqdm

--- a/VirtualBox/scripts/UPDATE.sh
+++ b/VirtualBox/scripts/UPDATE.sh
@@ -143,14 +143,18 @@ fi
 SIRF_INSTALL_PATH=$SIRF_SRC_PATH/install
 
 # best to use full path for python3/cython
-PYTHON_EXECUTABLE=`which python3`
-if [ -z "$PYTHON_EXECUTABLE" ]; then
-  PYTHON_EXECUTABLE=`which python`
+PYTHON_EXECUTABLE=$(which python3)
+if which python3; then
+  PYTHON_EXECUTABLE=$(which python3)
+else
+  PYTHON_EXECUTABLE=$(which python)
 fi
 PYTHON_EXECUTABLE=/usr/bin/python3
-CYTHON_EXECUTABLE=`which cython3`
-if [ -z "$CYTHON_EXECUTABLE" ]; then
-  CYTHON_EXECUTABLE=`which cython`
+
+if which cython3; then
+    CYTHON_EXECUTABLE=$(which cython3)
+else
+  CYTHON_EXECUTABLE=$(which cython)
 fi
 
 # ignore notebook keys, https://github.com/CCPPETMR/SIRF-Exercises/issues/20

--- a/VirtualBox/scripts/UPDATE.sh
+++ b/VirtualBox/scripts/UPDATE.sh
@@ -204,7 +204,7 @@ SuperBuild(){
         -DCYTHON_EXECUTABLE="$CYTHON_EXECUTABLE"\
         -DPYTHON_EXECUTABLE="$PYTHON_EXECUTABLE"\
         -DBUILD_pet_rd_tools=ON\
-        -DCYTHON_PATH="${HOME}/.local/bin"
+        -DCYTHON_ROOT="${HOME}/.local/bin"
   cmake --build . -j${num_parallel}
 
   if [ ! -f ${SIRF_INSTALL_PATH}/share/gadgetron/config/gadgetron.xml ]

--- a/VirtualBox/scripts/UPDATE.sh
+++ b/VirtualBox/scripts/UPDATE.sh
@@ -141,9 +141,17 @@ if [ -d $SIRF_SRC_PATH/CCPPETMR_VM ]; then
 fi
 
 SIRF_INSTALL_PATH=$SIRF_SRC_PATH/install
-# best to use full path
+
+# best to use full path for python3/cython
+PYTHON_EXECUTABLE=`which python3`
+if [ -z "$PYTHON_EXECUTABLE" ]; then
+  PYTHON_EXECUTABLE=`which python`
+fi
 PYTHON_EXECUTABLE=/usr/bin/python3
-CYTHON_EXECUTABLE=/usr/bin/cython3
+CYTHON_EXECUTABLE=`which cython3`
+if [ -z "$CYTHON_EXECUTABLE" ]; then
+  CYTHON_EXECUTABLE=`which cython`
+fi
 
 # ignore notebook keys, https://github.com/CCPPETMR/SIRF-Exercises/issues/20
 "$PYTHON_EXECUTABLE" -m pip install -U --user nbstripout

--- a/VirtualBox/scripts/UPDATE.sh
+++ b/VirtualBox/scripts/UPDATE.sh
@@ -203,7 +203,8 @@ SuperBuild(){
         -DBUILD_CIL=ON\
         -DCYTHON_EXECUTABLE="$CYTHON_EXECUTABLE"\
         -DPYTHON_EXECUTABLE="$PYTHON_EXECUTABLE"\
-        -DBUILD_pet_rd_tools=ON
+        -DBUILD_pet_rd_tools=ON\
+        -DCYTHON_PATH="${HOME}/.local/bin"
   cmake --build . -j${num_parallel}
 
   if [ ! -f ${SIRF_INSTALL_PATH}/share/gadgetron/config/gadgetron.xml ]

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -84,4 +84,4 @@ bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.
 
 chown -R $SIRFUSERNAME:users $userHOME
 export PATH=${userHOME}/.local/bin:${PATH}
-sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1
+sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -70,7 +70,7 @@ else
   # git pull
   git fetch -a
   git checkout VM_3.3.0
-  git pull --rebase origin VM_3.3.0
+  git pull origin VM_3.3.0
 fi
 
 

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -64,9 +64,13 @@ cd $userHOME/devel
 if [ ! -d $userHOME/devel/SIRF-SuperBuild ]; then
   git clone https://github.com/SyneRBI/SIRF-SuperBuild.git
   cd SIRF-SuperBuild
+  git checkout VM_3.3.0
 else
   cd SIRF-SuperBuild
   # git pull
+  git fetch -a
+  git checkout VM_3.3.0
+  git pull --rebase origin VM_3.3.0
 fi
 
 

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -79,4 +79,4 @@ bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.
 # Therefore no browser is installed on the VM
 
 chown -R $SIRFUSERNAME:users $userHOME
-sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh
+sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -64,13 +64,9 @@ cd $userHOME/devel
 if [ ! -d $userHOME/devel/SIRF-SuperBuild ]; then
   git clone https://github.com/SyneRBI/SIRF-SuperBuild.git
   cd SIRF-SuperBuild
-  git checkout VM_3.3.0
 else
   cd SIRF-SuperBuild
   # git pull
-  git fetch -a
-  git checkout VM_3.3.0
-  git pull origin VM_3.3.0
 fi
 
 

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -83,4 +83,5 @@ bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.
 # Therefore no browser is installed on the VM
 
 chown -R $SIRFUSERNAME:users $userHOME
+export PATH=${userHOME}/.local/bin:${PATH}
 sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -84,4 +84,4 @@ bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.
 
 chown -R $SIRFUSERNAME:users $userHOME
 export PATH=${userHOME}/.local/bin:${PATH}
-sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1
+# sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -84,4 +84,4 @@ bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.
 
 chown -R $SIRFUSERNAME:users $userHOME
 export PATH=${userHOME}/.local/bin:${PATH}
-# sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1
+sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1

--- a/VirtualBox/scripts/bootstrap_ubuntu.sh
+++ b/VirtualBox/scripts/bootstrap_ubuntu.sh
@@ -83,4 +83,4 @@ bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/INSTALL_python_packages.
 # Therefore no browser is installed on the VM
 
 chown -R $SIRFUSERNAME:users $userHOME
-sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0
+sudo -u $SIRFUSERNAME -H bash $userHOME/devel/SIRF-SuperBuild/VirtualBox/scripts/UPDATE.sh -t VM_3.3.0 -j 1

--- a/docker/build_essential-ubuntu.sh
+++ b/docker/build_essential-ubuntu.sh
@@ -22,17 +22,17 @@ pushd $INSTALL_DIR
 # CMake
 curl -o cmake.tgz -L https://github.com/Kitware/CMake/releases/download/v3.17.5/cmake-3.17.5-Linux-x86_64.tar.gz
 tar xzf cmake.tgz && rm cmake.tgz
-ln -s cmake-*x86_64 cmake
+ln -s cmake-*x86_64 cmake || true
 export PATH="$PWD/cmake/bin:$PATH"
 
 # ccache
 mkdir -p bin
 pushd bin
 # ccache compiler override
-ln -s "$(which ccache)" g++
-ln -s "$(which ccache)" g++-8
-ln -s "$(which ccache)" gcc
-ln -s "$(which ccache)" gcc-8
+ln -s "$(which ccache)" g++ || true
+ln -s "$(which ccache)" g++-8 || true
+ln -s "$(which ccache)" gcc || true
+ln -s "$(which ccache)" gcc-8 || true
 export PATH="$PWD:$PATH"
 popd
 

--- a/docker/requirements.txt
+++ b/docker/requirements.txt
@@ -1,11 +1,11 @@
 setuptools
 wheel
+numpy        # CIL
+scipy        # CIL
 docopt
 matplotlib
 git+https://github.com/ismrmrd/ismrmrd-python-tools.git@master#egg=ismrmrd-python-tools
 Cython       # CIL
-numpy        # CIL
-scipy        # CIL
 h5py         # CIL
 Pillow       # CIL
 wget         # CIL

--- a/docker/requirements.txt
+++ b/docker/requirements.txt
@@ -1,4 +1,5 @@
 setuptools
+wheel
 docopt
 matplotlib
 git+https://github.com/ismrmrd/ismrmrd-python-tools.git@master#egg=ismrmrd-python-tools

--- a/docker/requirements.txt
+++ b/docker/requirements.txt
@@ -16,3 +16,4 @@ tifffile
 nibabel
 deprecation
 nose
+setuptools

--- a/docker/requirements.txt
+++ b/docker/requirements.txt
@@ -1,3 +1,4 @@
+setuptools
 docopt
 matplotlib
 git+https://github.com/ismrmrd/ismrmrd-python-tools.git@master#egg=ismrmrd-python-tools
@@ -16,4 +17,3 @@ tifffile
 nibabel
 deprecation
 nose
-setuptools

--- a/docker/requirements.txt
+++ b/docker/requirements.txt
@@ -4,7 +4,7 @@ docopt
 matplotlib
 git+https://github.com/ismrmrd/ismrmrd-python-tools.git@master#egg=ismrmrd-python-tools
 Cython       # CIL
-numpy<=1.20  # CIL
+numpy        # CIL
 scipy        # CIL
 h5py         # CIL
 Pillow       # CIL

--- a/docker/requirements.txt
+++ b/docker/requirements.txt
@@ -12,7 +12,9 @@ wget         # CIL
 six          # CIL
 olefile      # CIL
 git+https://github.com/data-exchange/dxchange.git  # CIL
-numba        # CIL
+# https://numba.discourse.group/t/numba-0-50-0-and-llvmlite-0-33-0-released/19
+# https://github.com/numba/llvmlite/issues/621
+numba<0.51        # CIL
 pandas
 tifffile
 nibabel

